crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -439,15 +703,50 @@ where
         height: Option<Description::Height>,
         changes: Changes,
     ) -> StorageResult<()> {
-        let mut storage_transaction =
-            StorageTransaction::transaction(&self.db, ConflictPolicy::Overwrite, changes);
+        // cumulative_changes_lock_guard is defined to be Some only when the migration is in progress.
+        // If the migration is not in progress, the default set of changes will be used, and the overhead
+        // for handling caused by this function to handle the migration will be minimal.
+        let (cumulative_changes, maybe_cumulative_changes_lock_guard) =
+            self.take_migration_changes();
+
+        let mut migration_transaction = StorageTransaction::transaction(
+            &self.db,
+            ConflictPolicy::Overwrite,
+            // We need to keep ownership of the migration changes in case the current transaction fails.
+            // In this case we need to place the cumulative changes for migrating the modification history
+            // back into the lock guard.
+            // Note: this might be requiring cloning a fair amount of data
+            cumulative_changes.clone(),
+        );
+        let mut storage_transaction = StorageTransaction::transaction(
+            &mut migration_transaction,
+            ConflictPolicy::Overwrite,
+            changes,
+        );
+
         if let Some(height) = height {
             self.store_modifications_history(&mut storage_transaction, &height)?;
         }
 
-        self.db
-            .commit_changes(&storage_transaction.into_changes())?;
-        Ok(())
+        // This cannot fail
```

Comment:
This comment just caught my attention :)
How can we be sure it doesn't fail?

Reply:
Code digging :) 
This is because the transaction targets another transaction as the storage into which to commit, and the `ConflictPolicy` is specified to be `Overwrite`. 
When `.commit` is called, the `.commit_changes` function from the underlying inner storage is called (https://github.com/FuelLabs/fuel-core/blob/f5240cb8baa8210c7b744a49d96695df0c84b2f1/crates/storage/src/transactional.rs#L280). 
This is implemented here: https://github.com/FuelLabs/fuel-core/blob/f5240cb8baa8210c7b744a49d96695df0c84b2f1/crates/storage/src/transactional.rs#L287. If you follow the code, you will see that the function can fail only when `ConflictPolicy = Fail`. 

I'll clarify this in the comment

Reply:
Clarified in 4dc2643fc248d656ed772fb55efc8a77f2b9ca99